From a402c09d0fe7346b198f7d4c7fda96afb5497f83 Mon Sep 17 00:00:00 2001
From: OpenWrt Contributor <contributor@example.com>
Date: Mon, 28 Sep 2026 18:40:02 -0400
Subject: [PATCH 3/3] ath79: mikrotik: write NAND kernel images with
 kernel2minor

yafut relies on subpage writes for the final (partial) chunk of the
kernel file. Now that the AR934x driver disables subpage writes, yafut
fails with -EINVAL (NAND_NO_SUBPAGE_WRITE makes subpagesize equal to the
page size, so non-page-aligned writes are rejected):

  file_yaffs_write: error writing 'kernel': error -22 (Invalid argument)

Go back to building the NAND kernel image with kernel2minor and writing
it with nandwrite. This uses raw OOB data and does not depend on subpage
writes, which is what the ar71xx target already did and what RouterBOOT
v6 expects (ELF-in-YAFFS file named "kernel").

Signed-off-by: OpenWrt Contributor <contributor@example.com>
---
 include/image-commands.mk                     |  8 +++++
 target/linux/ath79/image/common-mikrotik.mk   |  8 ++---
 .../base-files/lib/upgrade/platform.sh        |  5 ++--
 target/linux/ath79/mikrotik/target.mk         |  2 +-
 tools/Makefile                                |  1 +
 tools/kernel2minor/Makefile                   | 29 +++++++++++++++++++
 6 files changed, 45 insertions(+), 8 deletions(-)
 create mode 100644 tools/kernel2minor/Makefile

diff --git a/include/image-commands.mk b/include/image-commands.mk
index 443c870..60a4e7f 100644
--- a/include/image-commands.mk
+++ b/include/image-commands.mk
@@ -27,6 +27,14 @@ define Build/append-kernel
 	dd if=$(IMAGE_KERNEL) >> $@
 endef
 
+define Build/kernel2minor
+	$(eval temp_file := $(shell mktemp))
+	cp $@ $(temp_file)
+	kernel2minor -k $(temp_file) -r $(temp_file).new $(1)
+	mv $(temp_file).new $@
+	rm -f $(temp_file)
+endef
+
 define Build/package-kernel-ubifs
 	mkdir $@.kernelubifs
 	cp $@ $@.kernelubifs/kernel
diff --git a/target/linux/ath79/image/common-mikrotik.mk b/target/linux/ath79/image/common-mikrotik.mk
index 1dfd515..3a93063 100644
--- a/target/linux/ath79/image/common-mikrotik.mk
+++ b/target/linux/ath79/image/common-mikrotik.mk
@@ -21,9 +21,7 @@ endef
 
 define Device/mikrotik_nand
   $(Device/mikrotik)
-  IMAGE/sysupgrade.bin = append-kernel | sysupgrade-tar | append-metadata
-  DEVICE_COMPAT_MESSAGE := \
-       NAND images switched to yafut. If running older image, reinstall from initramfs.
-  DEVICE_COMPAT_VERSION := 1.1
-
+  IMAGE/sysupgrade.bin = append-kernel | kernel2minor -s 2048 -e -c | \
+	sysupgrade-tar kernel=$$$$@ | append-metadata
+  DEVICE_PACKAGES := nand-utils
 endef
diff --git a/target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh b/target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh
index 67e8d77..fb41364 100644
--- a/target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh
@@ -50,7 +50,7 @@ platform_check_image() {
 	return 0
 }
 
-RAMFS_COPY_BIN='yafut'
+RAMFS_COPY_BIN='nandwrite'
 
 platform_do_upgrade_mikrotik_nand() {
 	CI_KERNPART=none
@@ -63,7 +63,8 @@ platform_do_upgrade_mikrotik_nand() {
 	board_dir=${board_dir%/}
 	[ -n "$board_dir" ] || return
 
-	tar xf "$1" ${board_dir}/kernel -O | yafut -d "$fw_mtd" -w -i - -o kernel -m 0755 || return
+	mtd erase kernel
+	tar xf "$1" ${board_dir}/kernel -O | nandwrite -o "$fw_mtd" -
 
 	nand_do_upgrade "$1"
 }
diff --git a/target/linux/ath79/mikrotik/target.mk b/target/linux/ath79/mikrotik/target.mk
index 0d314e8..7c14659 100644
--- a/target/linux/ath79/mikrotik/target.mk
+++ b/target/linux/ath79/mikrotik/target.mk
@@ -3,7 +3,7 @@ FEATURES += minor nand
 KERNELNAME := vmlinux vmlinuz
 IMAGES_DIR := ../../..
 
-DEFAULT_PACKAGES += kmod-ag71xx-legacy kmod-switch-ar8xxx wpad-basic-mbedtls yafut
+DEFAULT_PACKAGES += kmod-ag71xx-legacy kmod-switch-ar8xxx wpad-basic-mbedtls
 
 define Target/Description
 	Build firmware images for MikroTik devices based on Qualcomm Atheros
diff --git a/tools/Makefile b/tools/Makefile
index 79d3954..d1bbd0a 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -48,6 +48,7 @@ tools-y += firmware-utils
 tools-y += flex
 tools-y += gengetopt
 tools-y += gnulib
+tools-y += kernel2minor
 tools-y += libressl
 tools-y += libtool
 tools-y += lzma
diff --git a/tools/kernel2minor/Makefile b/tools/kernel2minor/Makefile
new file mode 100644
index 0000000..be730b4
--- /dev/null
+++ b/tools/kernel2minor/Makefile
@@ -0,0 +1,29 @@
+#
+# Copyright (C) 2016 adron@yapic.net
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=kernel2minor
+PKG_VERSION:=0.25
+PKG_RELEASE:=1
+
+PKG_SOURCE_URL:=https://github.com/adron-s/kernel2minor.git
+PKG_MIRROR_HASH:=6083c46c2fe0da37bacd04d5d5439c0e2a9d00e58ff47a63acfd5057d2aa2145
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_VERSION:=1e5a52c7941945f6d64807ebca4a5923ba5466bd
+PKG_HASH:=33ca413403a3341af0c9a8e6d9bb58f4ad080a5339e8a8729b83637d35bfaf1b
+
+include $(INCLUDE_DIR)/host-build.mk
+
+define Host/Install
+	$(INSTALL_BIN) $(HOST_BUILD_DIR)/kernel2minor $(STAGING_DIR_HOST)/bin/
+endef
+
+define Host/Clean
+	rm -f $(STAGING_DIR_HOST)/bin/kernel2minor
+endef
+
+$(eval $(call HostBuild))
-- 
2.55.0

